docs/proposals/failover/optimize-cluster-failover.md

```diff
@@ -0,0 +1,552 @@
+    - Migration method: Graceful, immediately
+    - Toleration time
+
+4. Eviction queue (The queue maintained in memory)
+- It is necessary to consider the overall health of the system and have the ability to downgrade eviction behavior and slow down queue processing.
+- During the eviction process, that is, when there are workloads to be evicted in the queue and the original faulty cluster recovers, the migration of these workloads can be abandoned.
+- Metrics for administrators need to be provided to facilitate problem location.
+- In the future, eviction based on workload priority, remaining resources, etc. can be considered, and advanced capabilities can be built relying on the eviction queue.
+
+5. Failover FeatureGate
+The Failover FeatureGate can impose overall constraints on the cluster failure migration mechanism. Once the Failover FeatureGate is turned off, the cluster failure judgment will no longer function, and no resources will be migrated.
+
+In light of the above design principles, it has been sorted out that the following modules in the current system do not meet the objectives of this proposal:
+
+1. In the judgment of cluster failure conditions, the "taint cluster by Conditions" rule is not provided for users to define.
+2. The design of several types of taints is incomplete and needs to be supplemented.
+3. There is no application-level failure migration strategy provided.
+4. The current eviction queue lacks the ability to be downgraded, and the queue does not provide metrics for administrators.
+5. The Failover FeatureGate cannot impose overall constraints on the cluster failure mechanism. Users can still trigger resource migration by adding taints.
+
+Next, let's introduce the design solutions for each of the above modules that do not meet the objectives one by one.
+
+### Cluster failure condition judgment
+
+Cluster failure condition judgment involves assessing the status conditions of a faulty cluster object to add various taints to the cluster. This process is referred to as "taint cluster by Conditions".
+
+This proposal adds a new API: TaintClusterByCondition, which can be used to describe what taints should be added to the target cluster when certain cluster status conditions are met. The relevant logic can be implemented by adding a new controller. This controller is responsible for monitoring changes in the cluster status conditions, that is, the `.status.conditions` field of the Cluster object, and then deciding how to modify the taints of the Cluster object according to the TaintClusterByCondition API defined by the user. When the matching conditions are no longer met, the taints added to the target cluster will be removed.
+
+The TaintClusterByCondition resource is a Cluster scope resource. Its API group is `policy.karmada.io` and the API version is `v1alpha1`.
+
+```go
+type TaintClusterByCondition struct {
```

Comment:
Agree with @RainbowMango about the resource name. `TaintClusterByCondition` sounds like an operation or function name, not a policy.

Reply:
How about `TaintClusterPolicy`, `ClusterTaintMatcher`, ...

Or do you have any suggestions?

Reply:
Update the API name to `ClusterTaintPolicy`.